tests/tx/scripts/test_sighash.py

```diff
@@ -0,0 +1,321 @@
+            weight=1,
+            inputs=[tokens_input],
+            outputs=[htr_output],
+            parents=parents,
+            tokens=[token_uid],
+            storage=self.manager1.tx_storage,
+            timestamp=token_creation_tx.timestamp + 1
+        )
+        self.manager1.cpu_mining_service.resolve(atomic_swap_tx)
+
+        # Alice signs her input using sighash bitmasks, instead of sighash_all.
+        # She also sets max inputs and max outputs limits, including one output for change.
+        sighash_bitmask = SighashBitmask(inputs=0b1, outputs=0b1)
+        data_to_sign1 = atomic_swap_tx.get_custom_sighash_data(sighash_bitmask)
+        assert self.manager1.wallet
+        public_bytes1, signature1 = self.manager1.wallet.get_input_aux_data(data_to_sign1, self.private_key1)
+        tokens_input.data = P2PKH.create_input_data(
+            public_key_bytes=public_bytes1,
+            signature=signature1,
+            sighash=sighash_bitmask,
+            inputs_outputs_limit=InputsOutputsLimit(max_inputs=2, max_outputs=3)
+        )
+
+        # At this point, the tx is partial. The inputs are valid, but they're mismatched with outputs
+        self.manager1.verification_service.verifiers.tx.verify_inputs(atomic_swap_tx)
+        with pytest.raises(InputOutputMismatch):
+            self.manager1.verification_service.verify(atomic_swap_tx)
+
+        # Alice sends the tx bytes to Bob, represented here by cloning the tx
+        atomic_swap_tx_clone = cast(Transaction, atomic_swap_tx.clone())
+        self.manager1.cpu_mining_service.resolve(atomic_swap_tx_clone)
+
+        # Bob creates an input spending all genesis HTR and adds it to the atomic swap tx
+        htr_input = TxInput(not_none(self.genesis_block.hash), 0, b'')
+        atomic_swap_tx_clone.inputs.append(htr_input)
+
+        # Bob adds an output to receive all test tokens
+        bob_output_script = P2PKH.create_output_script(self.address2)
+        tokens_output = TxOutput(token_creation_utxo.value, bob_output_script, 1)
+        atomic_swap_tx_clone.outputs.append(tokens_output)
+
+        # Bob adds two change outputs for his HTR, which violates the maximum tx outputs set by Alice
+        htr_output1 = TxOutput(int(genesis_utxo.value / 4), bob_output_script)
+        htr_output2 = TxOutput(int(genesis_utxo.value / 4), bob_output_script)
+        atomic_swap_tx_clone.outputs.append(htr_output1)
+        atomic_swap_tx_clone.outputs.append(htr_output2)
+
+        # Bob signs his input using sighash_all to complete the tx
+        data_to_sign2 = atomic_swap_tx_clone.get_sighash_all()
+        assert self.manager2.wallet
+        public_bytes2, signature2 = self.manager2.wallet.get_input_aux_data(data_to_sign2, self.genesis_private_key)
+        htr_input.data = P2PKH.create_input_data(public_bytes2, signature2)
+
+        # The atomic swap tx is not valid and cannot be propagated
+        with pytest.raises(InvalidInputData) as e:
```

Comment:
Should we create a specific exception for this case? It can inherit from `InvalidInputData`.

Reply:
My point is that this exception might be generated by another verification (instead of the one you are expecting to fail).